containers/src/Data/IntMap/Internal.hs

```diff
@@ -3475,69 +3499,82 @@ instance Read1 IntMap where
 {--------------------------------------------------------------------
   Link
 --------------------------------------------------------------------}
-link :: Prefix -> IntMap a -> Prefix -> IntMap a -> IntMap a
-link p1 t1 p2 t2 = linkWithMask (branchMask p1 p2) p1 t1 {-p2-} t2
+
+linkKey :: Key -> IntMap a -> Prefix -> IntMap a -> IntMap a
+linkKey k1 t1 (Prefix p2) t2 = link k1 t1 p2 t2
+{-# INLINE linkKey #-}
+
+link :: Int -> IntMap a -> Int -> IntMap a -> IntMap a
+link k1 t1 k2 t2 = linkWithMask (branchMask k1 k2) k1 t1 k2 t2
 {-# INLINE link #-}
 
 -- `linkWithMask` is useful when the `branchMask` has already been computed
-linkWithMask :: Mask -> Prefix -> IntMap a -> IntMap a -> IntMap a
-linkWithMask m p1 t1 {-p2-} t2
-  | zero p1 m = Bin p m t1 t2
-  | otherwise = Bin p m t2 t1
+linkWithMask :: Int -> Key -> IntMap a -> Key -> IntMap a -> IntMap a
+linkWithMask m k1 t1 k2 t2
+  | natFromInt k1 < natFromInt k2 = Bin p t1 t2
+  | otherwise = Bin p t2 t1
   where
-    p = mask p1 m
+    p = Prefix (mask k1 m + m)
```

Comment:
I think this addition could be a bitwise or, but that would need to be done before the sign conversion.

Reply:
> I think this addition could be a bitwise or

Yes. Do you suggest we use or?

> that would need to be done before the sign conversion.

I don't understand this part. Which sign conversion do you mean?

Reply:
Sign conversion as in `intFromNat` on `maskW`. Though actually yes, this is setting the lowest bit so the sign doesn't even matter.

> Do you suggest we use or?

In my mind it's a "cleaner" operation because there is no extra carry flag, so it's a "yes, but don't expect it to ever make any difference".

Reply:
That's understandable, updated to use bitwise or.
Note: There are more places in this file where bitwise or could be used (not introduced in this PR), but I'm not touching those for now.